### DISCUSSION TITLE
[TUF] How would timestamps be generated?

### DISCUSSION THREAD

**Discussion by sudo-bmitch** [notaryproject/notation#29] [Q&A]
With the current registry API, I don't believe there's any precedent for running code on the registry server to generate blobs. And a basic registry deployed without any features has no TLS encryption and allows anyone to push/pull from that service (admittedly that's not a model we want to encourage). For these reasons, I'm leaning against asking OCI to support running a timestamping service within the registry, however the TUF team is assuming otherwise. Is there an option to have an external timestamping service, and would that be more desirable than proposing a change to the OCI spec to support running a timestamp service within the registry?

One advantage to a separate timestamping service is allowing disconnect environments to run their own timestamping service separate from the upstream mirror. This service may have it's own keys, and clients in the disconnected environment may need to be configured to trust that key.

**Comment by mnm678** [comment]
With the TUF design, it should be possible to have the timestamp key controlled by a server outside of the registry that pushes signed timestamp files to the registry, given that this server is able to ensure they have the most recent snapshot. This is a little more circuitous than having it controlled by the registry itself, but would require fewer changes to registries. (cc @trishankatdatadog @justincappos in case I missed a security implication)

On a related note, I started a [document](https://github.com/notaryproject/nv2/pull/30) to track changes we'd require to the OCI spec to support TUF metadata.

**Comment by trishankatdatadog** [comment]
First of all, just to clarify: what do you mean by "timestamping" @sudo-bmitch? Do you mean as in TUF, or as in Windows codesigning?

**Reply by sudo-bmitch** [reply]
Timestamping as in the TUF timestamp signature on the snapshot.

**Comment by JustinCappos** [comment]
I don't see any problem with this given that a map file in TUF can point to
different keys for various roles.

**Comment by shizhMSFT** [comment]
For TUF, combining the notary service with the registry service is convenient but breaks the security model that the registry server is not trusted.

**Comment by JustinCappos** [comment]
It's more a question of "trusted for what"?  Trusted to know the latest
version of a piece of (snapshot) metadata is less impactful than being
trusted to provide the snapshot metadata or something more sensitive like
metadata pointing to images.
